File: trade_analysis_backend/app/schemas/future_fundamental_analysis.py
from datetime import datetime

from pydantic import ConfigDict, field_validator
from sqlmodel import SQLModel


class FutureFundamentalAnalysisBase(SQLModel):
    product_id: int
    report_id: int
    supply_side: str | None = None
    demand_side: str | None = None
    inventory_side: str | None = None
    industry_profit: str | None = None
    substitution_linkage: str | None = None
    policy_macro: str | None = None
    conclusion: str | None = None

    @field_validator(
        "supply_side",
        "demand_side",
        "inventory_side",
        "industry_profit",
        "substitution_linkage",
        "policy_macro",
        "conclusion",
        mode="before",
    )
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("field must be a string")
        cleaned = value.strip()
        return cleaned or None


class FutureFundamentalAnalysisCreate(FutureFundamentalAnalysisBase):
    pass


class FutureFundamentalAnalysisUpdate(SQLModel):
    analysis_id: int
    product_id: int | None = None
    report_id: int | None = None
    supply_side: str | None = None
    demand_side: str | None = None
    inventory_side: str | None = None
    industry_profit: str | None = None
    substitution_linkage: str | None = None
    policy_macro: str | None = None
    conclusion: str | None = None

    @field_validator(
        "supply_side",
        "demand_side",
        "inventory_side",
        "industry_profit",
        "substitution_linkage",
        "policy_macro",
        "conclusion",
        mode="before",
    )
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("field must be a string")
        cleaned = value.strip()
        return cleaned or None


class FutureFundamentalAnalysisDeleteRequest(SQLModel):
    analysis_id: int


class FutureFundamentalAnalysisRead(SQLModel):
    model_config = ConfigDict(from_attributes=True)

    analysis_id: int
    product_id: int
    report_id: int
    supply_side: str | None
    demand_side: str | None
    inventory_side: str | None
    industry_profit: str | None
    substitution_linkage: str | None
    policy_macro: str | None
    conclusion: str | None
    create_at: datetime
    updated_at: datetime


class FutureFundamentalAnalysisListItem(FutureFundamentalAnalysisRead):
    product_code: str
    product_display_name: str
    report_name: str
    report_storage_path: str

File: trade_analysis_web/src/api/modules/futureFundamentalAnalysis.ts
import axios from "@/api/axios";

export interface FutureReportDocument {
  report_id: number
  report_name: string
  published_at: string
  report_source: string
  storage_path: string
  original_filename: string
  content_type: string
  file_size: number
  create_at: string
  updated_at: string
}

export interface FutureReportUploadResult extends FutureReportDocument {
  url: string
}

export interface FutureFundamentalAnalysis {
  analysis_id: number
  product_id: number
  report_id: number
  supply_side: string | null
  demand_side: string | null
  inventory_side: string | null
  industry_profit: string | null
  substitution_linkage: string | null
  policy_macro: string | null
  conclusion: string | null
  create_at: string
  updated_at: string
  product_code: string
  product_display_name: string
  report_name: string
  report_storage_path: string
}

export interface FutureFundamentalAnalysisCreateParams {
  product_id: number
  report_id: number
  supply_side?: string | null
  demand_side?: string | null
  inventory_side?: string | null
  industry_profit?: string | null
  substitution_linkage?: string | null
  policy_macro?: string | null
  conclusion?: string | null
}

export interface FutureFundamentalAnalysisUpdateParams
  extends Partial<FutureFundamentalAnalysisCreateParams> {
  analysis_id: number
}

export const getFutureReportDocumentListApi = () => {
  return axios.get<FutureReportDocument[]>("/future-report-documents") as unknown as Promise<FutureReportDocument[]>
}

export const uploadFutureReportDocumentApi = (params: {
  published_at: string
  report_source: string
  file: File
}) => {
  const formData = new FormData()
  formData.append("published_at", params.published_at)
  formData.append("report_source", params.report_source)
  formData.append("file", params.file)
  return axios.post<FutureReportUploadResult>("/future-report-documents/upload", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  }) as unknown as Promise<FutureReportUploadResult>
}

export const deleteFutureReportDocumentApi = (reportId: number) => {
  return axios.post<void>("/future-report-documents/delete", { report_id: reportId }) as unknown as Promise<void>
}

export const getFutureFundamentalAnalysisListApi = () => {
  return axios.get<FutureFundamentalAnalysis[]>("/future-fundamental-analyses") as unknown as Promise<
    FutureFundamentalAnalysis[]
  >
}

export const createFutureFundamentalAnalysisApi = (params: FutureFundamentalAnalysisCreateParams) => {
  return axios.post("/future-fundamental-analyses/create", params) as unknown as Promise<void>
}

export const updateFutureFundamentalAnalysisApi = (params: FutureFundamentalAnalysisUpdateParams) => {
  return axios.post("/future-fundamental-analyses/update", params) as unknown as Promise<void>
}

export const deleteFutureFundamentalAnalysisApi = (analysisId: number) => {
  return axios.post<void>("/future-fundamental-analyses/delete", {
    analysis_id: analysisId,
  }) as unknown as Promise<void>
}

export const resolveFutureReportPdfUrl = (path?: string | null) => {
  if (!path) {
    return ""
  }
  if (/^https?:\/\//.test(path)) {
    return path
  }
  return `/storage/${path.replace(/^\/+/, "")}`
}
